MS1 gets mismatched events in assert task

## Thanks for filing a bug report :-)

This seems not to be a problem in our code just that SP might have missed an event.

## Steps to Reproduce
1. 
2. 

## Expected Result
- 

## Actual Result
```
{
  "task": "<AssertBlockchainEventsTask: {'contract_name': 'TokenNetwork', 'event_name': 'NonClosingBalanceProofUpdated', 'num_events': 1, 'event_args': {'closing_participant': 0}}>",
  "event": "Task errored",
  "logger": "scenario_player.tasks.base",
  "level": "error",
  "greenlet_name": "orchestration",
  "timestamp": "2021-01-05 00:30:31.752913",
  "exception": "Traceback (most recent call last):\n  File \"/usr/local/lib/python3.7/site-packages/scenario_player/tasks/base.py\", line 134, in __call__\n    sleep(1)\n  File \"/usr/local/lib/python3.7/site-packages/gevent/hub.py\", line 163, in sleep\n    hub.wait(t)\n  File \"src/gevent/_hub_primitives.py\", line 46, in gevent._gevent_c_hub_primitives.WaitOperationsGreenlet.wait\n  File \"src/gevent/_hub_primitives.py\", line 55, in gevent._gevent_c_hub_primitives.WaitOperationsGreenlet.wait\n  File \"src/gevent/_waiter.py\", line 151, in gevent._gevent_c_waiter.Waiter.get\n  File \"src/gevent/_greenlet_primitives.py\", line 61, in gevent._gevent_c_greenlet_primitives.SwitchOutGreenletWithLoop.switch\n  File \"src/gevent/_greenlet_primitives.py\", line 61, in gevent._gevent_c_greenlet_primitives.SwitchOutGreenletWithLoop.switch\n  File \"src/gevent/_greenlet_primitives.py\", line 65, in gevent._gevent_c_greenlet_primitives.SwitchOutGreenletWithLoop.switch\n  File \"src/gevent/_gevent_c_greenlet_primitives.pxd\", line 35, in gevent._gevent_c_greenlet_primitives._greenlet_switch\ngevent.timeout.Timeout: 300 seconds\n\nDuring handling of the above exception, another exception occurred:\n\nTraceback (most recent call last):\n  File \"/usr/local/lib/python3.7/site-packages/scenario_player/tasks/base.py\", line 138, in __call__\n    raise exception\n  File \"/usr/local/lib/python3.7/site-packages/scenario_player/tasks/base.py\", line 126, in __call__\n    return_val = self._run(*args, **kwargs)\n  File \"/usr/local/lib/python3.7/site-packages/scenario_player/tasks/blockchain.py\", line 185, in _run\n    f\"Expected number of events ({self.num_events}) did not match the number \"\nscenario_player.exceptions.ScenarioAssertionError: Expected number of events (1) did not match the number of events found (0)"
}
```

Fails at assert task
```
AssertBlockchainEventsTask: {'contract_name': 'TokenNetwork', 'event_name': 'NonClosingBalanceProofUpdated', 'num_events': 1, 'event_args': {'closing_participant': 0}}
```

```
"Traceback (most recent call last):
File \"/usr/local/lib/python3.7/site-packages/scenario_player/tasks/base.py\", line 134, in __call__
sleep(1)
File \"/usr/local/lib/python3.7/site-packages/gevent/hub.py\", line 163, in sleep
hub.wait(t)
File \"src/gevent/_hub_primitives.py\", line 46, in gevent._gevent_c_hub_primitives.WaitOperationsGreenlet.wait
File \"src/gevent/_hub_primitives.py\", line 55, in gevent._gevent_c_hub_primitives.WaitOperationsGreenlet.wait
File \"src/gevent/_waiter.py\", line 151, in gevent._gevent_c_waiter.Waiter.get
File \"src/gevent/_greenlet_primitives.py\", line 61, in gevent._gevent_c_greenlet_primitives.SwitchOutGreenletWithLoop.switch
File \"src/gevent/_greenlet_primitives.py\", line 61, in gevent._gevent_c_greenlet_primitives.SwitchOutGreenletWithLoop.switch
File \"src/gevent/_greenlet_primitives.py\", line 65, in gevent._gevent_c_greenlet_primitives.SwitchOutGreenletWithLoop.switch
File \"src/gevent/_gevent_c_greenlet_primitives.pxd\", line 35, in gevent._gevent_c_greenlet_primitives._greenlet_switch
gevent.timeout.Timeout: 300 seconds

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
File \"/usr/local/lib/python3.7/site-packages/scenario_player/tasks/base.py\", line 138, in __call__
raise exception
File \"/usr/local/lib/python3.7/site-packages/scenario_player/tasks/base.py\", line 126, in __call__
return_val = self._run(*args, **kwargs)
File \"/usr/local/lib/python3.7/site-packages/scenario_player/tasks/blockchain.py\", line 185, in _run
f\"Expected number of events ({self.num_events}) did not match the number \"
scenario_player.exceptions.ScenarioAssertionError: Expected number of events (1) did not match the number of events found (0)"
```

Just an observation from node0 logs 
```
{ blockNumber: 4048915 } }
2021-01-05T00:30:31.765Z [info] Stopping raiden...
2021-01-05T00:30:31.767Z [info] Closing server...
2021-01-05T00:30:31.769Z [info] %c action     color: #03A9F4; font-weight: bold { type: 'raiden/shutdown', payload: { reason: 'raidenStopped' } }
2021-01-05T00:30:31.769Z [info] %c action     color: #03A9F4; font-weight: bold {
  type: 'message/send/success',
  payload: { via: '!JoAzRQkqJBnsFswyFU:transport.transport01.raiden.network' },
  meta: {
    address: '0x786D5A4B39e1065d2Bfd8b9B7DB253FEd8dbF607',
    msgId: '1609805616209'
  }
}
```

`message success` action after `raidenStopped` though this might just be due to asynchronous nature of all the events.

[scenario-player-run_ms1_simple_monitoring_2021-01-05T00:08:08.log](https://github.com/raiden-network/light-client/files/5769816/scenario-player-run_ms1_simple_monitoring_2021-01-05T00.08.08.log)


[ms1_node0_node1_5thJan2021.tar.gz](https://github.com/raiden-network/light-client/files/5769823/ms1_node0_node1_5thJan2021.tar.gz)

--- a/raiden-cli/CHANGELOG.md
+++ b/raiden-cli/CHANGELOG.md
@@ -4,7 +4,11 @@
 ### Added
 - [#2417] Register sensible endpoints on 'synced' SDK promise, updates /status API accordingly
 
+### Changed
+- [#2505] Wait for background tasks to gracefully complete before exiting
+
 [#2417]: https://github.com/raiden-network/light-client/pull/2417
+[#2505]: https://github.com/raiden-network/light-client/pull/2505
 
 ## [0.14.0] - 2020-11-25
 ### Fixed

--- a/raiden-cli/src/index.ts
+++ b/raiden-cli/src/index.ts
@@ -215,13 +215,6 @@ function shutdownServer(this: Cli): void {
     this.log.info('Closing server...');
     this.server.close();
   }
-  // it's safe to call stop multiple times; this one to await for db flush;
-  // then we wait a little for any pending request to complete and force-exit cleanly after that,
-  // to skip wrtc segfault on process teardown
-  new Promise((resolve) => setTimeout(resolve, 10))
-    .then(() => this.raiden.stop())
-    .then(() => new Promise((resolve) => setTimeout(resolve, 1000)))
-    .then(() => process.exit(0));
 }
 
 function shutdownRaiden(this: Cli): void {
@@ -241,6 +234,13 @@ function registerShutdownHooks(this: Cli): void {
   });
   process.on('SIGINT', shutdownRaiden.bind(this));
   process.on('SIGTERM', shutdownRaiden.bind(this));
+  // 'beforeExit' is emitted after all background tasks are finished;
+  // we need to call process.exit explicitly in order to avoid 'wrtc'
+  // cleanup from segfaulting the process
+  process.on('beforeExit', (code) => {
+    this.log.info('Exiting', code);
+    process.exit(code);
+  });
 }
 
 async function createRaidenConfig(

--- a/raiden-ts/CHANGELOG.md
+++ b/raiden-ts/CHANGELOG.md
@@ -9,6 +9,7 @@
 
 ### Changed
 - [#2409] Lower default payment expiration to 1.1 × reveal timeout
+- [#2505] Properly shut down epics on stop and wait for teardown/cleanup tasks
 
 ### Fixed
 - [#2352] Presence bug, transport fixes and performance improvements
@@ -19,6 +20,7 @@
 [#2417]: https://github.com/raiden-network/light-client/pull/2417
 [#2444]: https://github.com/raiden-network/light-client/issues/2444
 [#2446]: https://github.com/raiden-network/light-client/issues/2446
+[#2505]: https://github.com/raiden-network/light-client/pull/2505
 
 ## [0.14.0] - 2020-11-25
 ### Fixed

--- a/raiden-ts/src/channels/actions.ts
+++ b/raiden-ts/src/channels/actions.ts
@@ -15,6 +15,10 @@ const ChannelId = t.type({
 export const newBlock = createAction('block/new', t.type({ blockNumber: t.number }));
 export interface newBlock extends ActionType<typeof newBlock> {}
 
+/* A new blockTime (average time between latest X blocks) was detected */
+export const blockTime = createAction('block/time', t.type({ blockTime: t.number }));
+export interface blockTime extends ActionType<typeof blockTime> {}
+
 /**
  * A new token network is detected in the TokenNetworkRegistry instance
  * fromBlock is only set on the first time, to fetch and handle past events

--- a/raiden-ts/src/channels/epics.ts
+++ b/raiden-ts/src/channels/epics.ts
@@ -62,7 +62,14 @@ import {
 import { chooseOnchainAccount, getContractWithSigner } from '../helpers';
 import { Address, Hash, UInt, Signature, isntNil, HexString, last } from '../utils/types';
 import { isActionOf } from '../utils/actions';
-import { pluckDistinct, distinctRecordValues, retryAsync$, takeIf, retryWhile } from '../utils/rx';
+import {
+  pluckDistinct,
+  distinctRecordValues,
+  retryAsync$,
+  takeIf,
+  retryWhile,
+  completeWith,
+} from '../utils/rx';
 import { fromEthersEvent, getLogsByChunk$, logToContractEvent } from '../utils/ethers';
 import { encode } from '../utils/data';
 
@@ -133,13 +140,14 @@ export function initEpic(
  * @returns Observable of newBlock actions
  */
 export function initNewBlockEpic(
-  {}: Observable<RaidenAction>,
+  action$: Observable<RaidenAction>,
   {}: Observable<RaidenState>,
   { provider }: RaidenEpicDeps,
 ): Observable<newBlock> {
   return retryAsync$(() => provider.getBlockNumber(), provider.pollingInterval).pipe(
     mergeMap((blockNumber) => merge(of(blockNumber), fromEthersEvent<number>(provider, 'block'))),
     map((blockNumber) => newBlock({ blockNumber })),
+    completeWith(action$),
   );
 }
 
@@ -232,13 +240,14 @@ export function initTokensRegistryEpic(
  * @returns Observable of raidenShutdown actions
  */
 export function initMonitorProviderEpic(
-  {}: Observable<RaidenAction>,
+  action$: Observable<RaidenAction>,
   {}: Observable<RaidenState>,
   { address, main, network, provider }: RaidenEpicDeps,
 ): Observable<raidenShutdown> {
   const mainAddress = main?.address ?? address;
   let isProviderAccount: boolean | undefined;
   return timer(0, provider.pollingInterval).pipe(
+    completeWith(action$),
     exhaustMap(async () => {
       try {
         const [accounts, currentNetwork] = await Promise.all([
@@ -601,6 +610,7 @@ export function channelEventsEpic(
         }),
       ),
     ),
+    completeWith(action$),
   );
 }
 
@@ -1479,7 +1489,7 @@ export function confirmationEpic(
   return combineLatest([
     state$.pipe(pluckDistinct('blockNumber')),
     state$.pipe(pluck('pendingTxs')),
-    config$.pipe(pluckDistinct('confirmationBlocks')),
+    config$.pipe(pluckDistinct('confirmationBlocks'), completeWith(state$)),
   ]).pipe(
     filter(([, pendingTxs]) => pendingTxs.length > 0),
     // exhaust will ignore blocks while concat$ is busy

--- a/raiden-ts/src/channels/epics.ts
+++ b/raiden-ts/src/channels/epics.ts
@@ -38,6 +38,7 @@ import {
 import sortBy from 'lodash/sortBy';
 import isEmpty from 'lodash/isEmpty';
 import findKey from 'lodash/findKey';
+import constant from 'lodash/constant';
 
 import type { BigNumber } from '@ethersproject/bignumber';
 import { Zero } from '@ethersproject/constants';
@@ -90,6 +91,7 @@ import {
   channelSettle,
   channelSettleable,
   channelWithdrawn,
+  blockTime,
 } from './actions';
 import { assertTx, channelKey, groupChannel$, channelUniqueKey, approveIfNeeded$ } from './utils';
 
@@ -148,6 +150,54 @@ export function initNewBlockEpic(
     mergeMap((blockNumber) => merge(of(blockNumber), fromEthersEvent<number>(provider, 'block'))),
     map((blockNumber) => newBlock({ blockNumber })),
     completeWith(action$),
+  );
+}
+
+/**
+ * Every fetchEach=20 blocks, update average block time since previous request
+ *
+ * @param action$ - Observable of RaidenActions
+ * @param state$ - Observable of RaidenStates
+ * @param deps - RaidenEpicDeps members
+ * @param deps.provider - Eth provider
+ * @returns Observable of blockTime actions
+ */
+export function blockTimeEpic(
+  action$: Observable<RaidenAction>,
+  {}: Observable<RaidenState>,
+  { provider }: RaidenEpicDeps,
+): Observable<blockTime> {
+  const fetchEach = 20;
+  type BlockInfo = readonly [blockNumber: number, timestamp: number, blockTime?: number];
+  let lastInfo: BlockInfo = [-fetchEach, 0]; // previously fetched block info
+
+  // get block info for a given block number
+  const getBlockInfo$ = (blockNumber: number) =>
+    defer(async () =>
+      provider
+        .getBlock(blockNumber)
+        .then(({ timestamp }): BlockInfo => [blockNumber, timestamp * 1000]),
+    );
+
+  return action$.pipe(
+    filter(newBlock.is),
+    pluck('payload', 'blockNumber'),
+    filter((blockNumber) => blockNumber >= lastInfo[0] + fetchEach),
+    exhaustMap((blockNumber) => {
+      const prevInfo$ =
+        lastInfo[0] > 0 ? of(lastInfo) : getBlockInfo$(Math.max(1, blockNumber - fetchEach));
+      const curInfo$ = getBlockInfo$(blockNumber);
+      return combineLatest([prevInfo$, curInfo$]).pipe(
+        mergeMap(function* ([prevInfo, curInfo]) {
+          const avgBlockTime = (curInfo[1] - prevInfo[1]) / (curInfo[0] - prevInfo[0]);
+          // emit a new avg blockTime only if it changed
+          if (avgBlockTime !== lastInfo[2]) yield avgBlockTime;
+          lastInfo = [curInfo[0], curInfo[1], avgBlockTime]; // persist last BlockInfo
+        }),
+        catchError(constant(EMPTY)), // ignore errors to retry next block
+      );
+    }),
+    map((avgBlockTime) => blockTime({ blockTime: avgBlockTime })),
   );
 }
 

--- a/raiden-ts/src/db/epics.ts
+++ b/raiden-ts/src/db/epics.ts
@@ -1,12 +1,13 @@
 import { Observable } from 'rxjs';
-import { ignoreElements, first, finalize } from 'rxjs/operators';
+import { ignoreElements } from 'rxjs/operators';
 
 import { RaidenAction } from '../actions';
 import { RaidenState } from '../state';
 import { RaidenEpicDeps } from '../types';
+import { completeWith } from '../utils/rx';
 
 /**
- * Shutdown database instance when raiden shuts down
+ * An epic to error globally in case db.busy$ errors (i.e. database errors)
  *
  * @param action$ - Observable of RaidenActions
  * @param state$ - Observable of RaidenStates
@@ -14,22 +15,10 @@ import { RaidenEpicDeps } from '../types';
  * @param deps.db - Database instance
  * @returns observable to shutdown db instance on raidenShutdown
  */
-export function dbShutdownEpic(
+export function dbErrorsEpic(
   action$: Observable<RaidenAction>,
   {}: Observable<RaidenState>,
   { db }: RaidenEpicDeps,
 ): Observable<never> {
-  return action$.pipe(
-    ignoreElements(),
-    finalize(async () => {
-      await db.busy$.pipe(first((busy) => !busy)).toPromise();
-      db.busy$.next(true);
-      try {
-        await db.close();
-      } finally {
-        db.busy$.next(false);
-        db.busy$.complete();
-      }
-    }),
-  );
+  return db.busy$.pipe(ignoreElements(), completeWith(action$));
 }

--- a/raiden-ts/src/epics.ts
+++ b/raiden-ts/src/epics.ts
@@ -1,16 +1,4 @@
-import {
-  Observable,
-  from,
-  of,
-  combineLatest,
-  using,
-  throwError,
-  merge,
-  timer,
-  defer,
-  EMPTY,
-  concat,
-} from 'rxjs';
+import { Observable, from, of, combineLatest, using, timer, defer, EMPTY, concat } from 'rxjs';
 import {
   catchError,
   filter,
@@ -29,6 +17,11 @@ import {
   endWith,
   shareReplay,
   withLatestFrom,
+  delayWhen,
+  take,
+  first,
+  tap,
+  finalize,
 } from 'rxjs/operators';
 import { MaxUint256 } from '@ethersproject/constants';
 import negate from 'lodash/negate';
@@ -46,7 +39,6 @@ import { getPresences$ } from './transport/utils';
 import { rtcChannel } from './transport/actions';
 import { pfsListUpdated, udcDeposit } from './services/actions';
 import { Address, UInt } from './utils/types';
-import { isActionOf } from './utils/actions';
 
 import * as DatabaseEpics from './db/epics';
 import * as ChannelsEpics from './channels/epics';
@@ -274,36 +266,59 @@ export function raidenRootEpic(
   deps: RaidenEpicDeps,
 ): Observable<RaidenAction> {
   // observable which emits once when a raidenShutdown action goes through actions pipeline
-  const shutdownNotification = action$.pipe(filter(isActionOf(raidenShutdown))),
+  const shutdownNotification = action$.pipe(filter(raidenShutdown.is), take(1)),
     // actions pipeline, but ends with (including) a raidenShutdown action
-    limitedAction$ = action$.pipe(
-      takeWhile<RaidenAction>(negate(isActionOf(raidenShutdown)), true),
-    ),
+    limitedAction$ = action$.pipe(takeWhile<RaidenAction>(negate(raidenShutdown.is), true)),
     // states pipeline, but ends when shutdownNotification emits
     limitedState$ = state$.pipe(takeUntil(shutdownNotification));
 
+  // like combineEpics, but completes action$, state$ & output$ when a raidenShutdown goes through;
   return using(
     // wire deps.latest$ when observableFactory below gets subscribed, and tears down on complete
     () => {
-      const sub = getLatest$(limitedAction$, limitedState$, deps).subscribe(deps.latest$);
+      const sub = getLatest$(action$, state$, deps).subscribe(deps.latest$);
       // ensure deps.latest$ is completed if teardown happens before getLatest$ completion
       sub.add(() => deps.latest$.complete());
       return sub;
     },
-    () =>
-      // like combineEpics, but completes action$, state$ & output$ when a raidenShutdown goes through;
-      // also merge deps.db.busy$ errors, in order to shut down epics if they occur
-      merge(from(Object.values(RaidenEpics)), deps.db.busy$.pipe(ignoreElements())).pipe(
-        mergeMap((epic) =>
-          epic(limitedAction$, limitedState$, deps).pipe(
+    () => {
+      const subscribedEpics = new Set<string>();
+      // main epics output
+      const output$ = from(Object.values(RaidenEpics)).pipe(
+        mergeMap((epic) => {
+          subscribedEpics.add(epic.name);
+          return epic(limitedAction$, limitedState$, deps).pipe(
             catchError((err) => {
               deps.log.error('Epic error', epic.name, epic, err);
-              return throwError(err);
+              return of(raidenShutdown({ reason: err }));
             }),
+            finalize(() => subscribedEpics.delete(epic.name)),
+          );
+        }),
+        takeUntil(
+          shutdownNotification.pipe(
+            withLatestFrom(deps.config$),
+            // give up to httpTimeout for the epics to complete on their own
+            delayWhen(([_, { httpTimeout }]) => timer(httpTimeout)),
+            tap(() => deps.log.warn('Pending Epics :', subscribedEpics)),
           ),
         ),
-        catchError((err) => of(raidenShutdown({ reason: err }))),
-        takeUntil(shutdownNotification),
-      ),
+      );
+      // also concat db teardown tasks, to be done after main epic completes
+      const teardown$ = deps.db.busy$.pipe(
+        first((busy) => !busy),
+        tap(() => deps.db.busy$.next(true)),
+        // ignore db.busy$ errors, they're merged in the output by dbErrorsEpic
+        catchError(() => of(null)),
+        mergeMap(async () => deps.db.close()),
+        ignoreElements(),
+        finalize(() => {
+          deps.db.busy$.next(false);
+          deps.db.busy$.complete();
+        }),
+      );
+      // subscribe to teardown$ only after output$ completes
+      return concat(output$, teardown$);
+    },
   );
 }

--- a/raiden-ts/src/epics.ts
+++ b/raiden-ts/src/epics.ts
@@ -1,4 +1,4 @@
-import { Observable, from, of, combineLatest, using, timer, defer, EMPTY, concat } from 'rxjs';
+import { Observable, from, of, combineLatest, using, timer, concat } from 'rxjs';
 import {
   catchError,
   filter,
@@ -12,7 +12,6 @@ import {
   distinctUntilChanged,
   ignoreElements,
   switchMap,
-  exhaustMap,
   mapTo,
   endWith,
   shareReplay,
@@ -27,7 +26,6 @@ import { MaxUint256 } from '@ethersproject/constants';
 import negate from 'lodash/negate';
 import unset from 'lodash/fp/unset';
 import isEqual from 'lodash/isEqual';
-import constant from 'lodash/constant';
 
 import { RaidenState } from './state';
 import { RaidenEpicDeps, Latest } from './types';
@@ -45,58 +43,7 @@ import * as ChannelsEpics from './channels/epics';
 import * as TransportEpics from './transport/epics';
 import * as TransfersEpics from './transfers/epics';
 import * as ServicesEpics from './services/epics';
-
-/**
- * Returns a shared observable to calculate average block time every [opts.fetchEach] blocks
- *
- * @param state$ - Observable of RaidenStates, to fetch blockNumber updates
- * @param provider - To fetch historic block info
- * @param opts - Options object
- * @param opts.initialBlockTime - constant for an initial blockTime (at subscription time)
- * @param opts.fetchEach - how often to fetch block data (in blocks)
- * @returns average block time, in milliseconds, from latest [opts.fetchEach] blocks
- */
-function getBlockTime$(
-  state$: Observable<RaidenState>,
-  provider: RaidenEpicDeps['provider'],
-  { initialBlockTime, fetchEach }: { initialBlockTime: number; fetchEach: number } = {
-    initialBlockTime: 15e3,
-    fetchEach: 10,
-  },
-): Observable<number> {
-  type BlockInfo = readonly [blockNumber: number, timestamp: number, blockTime?: number];
-  let lastInfo: BlockInfo = [-fetchEach, 0]; // previously fetched block info
-
-  // get block info for a given block number
-  const getBlockInfo$ = (blockNumber: number) =>
-    defer(async () =>
-      provider
-        .getBlock(blockNumber)
-        .then((block): BlockInfo => [blockNumber, block.timestamp * 1000]),
-    );
-
-  return concat(
-    // like startWith, but if somehow this gets re-subscribed, uses lastInfo[2] as initialBlockTime
-    defer(() => (lastInfo[0] > 0 && lastInfo[2] ? of(lastInfo[2]) : of(initialBlockTime))),
-    state$.pipe(
-      pluckDistinct('blockNumber'),
-      filter((blockNumber) => blockNumber >= lastInfo[0] + fetchEach),
-      exhaustMap((blockNumber) => {
-        const prevInfo$ =
-          lastInfo[0] > 0 ? of(lastInfo) : getBlockInfo$(Math.max(1, blockNumber - fetchEach));
-        const curInfo$ = getBlockInfo$(blockNumber);
-        return combineLatest([prevInfo$, curInfo$]).pipe(
-          map(([prevInfo, curInfo]) => {
-            const avgBlockTime = (curInfo[1] - prevInfo[1]) / (curInfo[0] - prevInfo[0]);
-            lastInfo = [curInfo[0], curInfo[1], avgBlockTime];
-            return avgBlockTime;
-          }),
-          catchError(constant(EMPTY)), // ignore errors to retry next block
-        );
-      }),
-    ),
-  ).pipe(shareReplay(1)); // share observable to reuse subject in case of multiple subscriptions
-}
+import { blockTime } from './channels/actions';
 
 // Observable of truthy if stale (no new blocks for too long, possibly indicating eth node is out
 // of sync), falsy otherwise/initially
@@ -192,7 +139,12 @@ export function getLatest$(
     startWith(MaxUint256 as UInt<32>),
     distinctUntilChanged((a, b) => a.eq(b)),
   );
-  const blockTime$ = getBlockTime$(state$, provider);
+  const blockTime$ = action$.pipe(
+    filter(blockTime.is),
+    pluck('payload', 'blockTime'),
+    startWith(15e3), // default initial blockTime of 15s
+    shareReplay(1),
+  );
   const config$ = getConfig$(defaultConfig, state$, { udcBalance$, blockTime$ });
   const presences$ = getPresences$(action$);
   const pfsList$ = action$.pipe(

--- a/raiden-ts/src/services/epics.ts
+++ b/raiden-ts/src/services/epics.ts
@@ -61,7 +61,7 @@ import { isActionOf, isResponseOf } from '../utils/actions';
 import { encode, jsonParse, jsonStringify } from '../utils/data';
 import { fromEthersEvent, logToContractEvent } from '../utils/ethers';
 import { RaidenError, ErrorCodes, assert, networkErrors, commonTxErrors } from '../utils/error';
-import { pluckDistinct, retryAsync$, retryWhile } from '../utils/rx';
+import { completeWith, pluckDistinct, retryAsync$, retryWhile } from '../utils/rx';
 import { matrixPresence } from '../transport/actions';
 import { getCap } from '../transport/utils';
 import type { UserDeposit, HumanStandardToken } from '../contracts';
@@ -373,7 +373,7 @@ export function pfsFeeUpdateEpic(
  * @returns Observable of pfsListUpdated actions
  */
 export function pfsServiceRegistryMonitorEpic(
-  {}: Observable<RaidenAction>,
+  action$: Observable<RaidenAction>,
   {}: Observable<RaidenState>,
   { provider, serviceRegistryContract, contractsInfo, config$, latest$, init$ }: RaidenEpicDeps,
 ): Observable<pfsListUpdated> {
@@ -443,6 +443,7 @@ export function pfsServiceRegistryMonitorEpic(
           }),
         );
     }),
+    completeWith(action$),
   );
 }
 
@@ -717,11 +718,10 @@ function makeMonitoringRequest$({
  */
 export function monitorRequestEpic(
   {}: Observable<RaidenAction>,
-  {}: Observable<RaidenState>,
+  state$: Observable<RaidenState>,
   deps: RaidenEpicDeps,
 ): Observable<messageGlobalSend> {
-  return deps.latest$.pipe(
-    pluck('state'),
+  return state$.pipe(
     groupChannel$,
     withLatestFrom(deps.config$),
     mergeMap(([grouped$, { httpTimeout }]) =>
@@ -894,7 +894,8 @@ export function udcWithdrawPlannedEpic(
     mergeMap((action) =>
       state$.pipe(
         pluck('blockNumber'),
-        first((blockNumber) => action.payload.block < blockNumber),
+        filter((blockNumber) => action.payload.block < blockNumber),
+        take(1),
         mapTo(action),
       ),
     ),
@@ -929,7 +930,8 @@ export function udcWithdrawPlannedEpic(
                 onErrors: networkErrors,
               }),
             ),
-            first((newBalance) => newBalance.lt(balance)),
+            filter((newBalance) => newBalance.lt(balance)),
+            take(1),
             map((newBalance) =>
               udcWithdrawn(
                 {
@@ -991,6 +993,7 @@ export function msMonitorNewBPEpic(
         confirmationBlocks,
       ),
     ),
+    completeWith(state$),
     map(
       logToContractEvent<[Address, UInt<32>, UInt<32>, UInt<8>, Address, Address, Event]>(
         monitoringServiceContract,

--- a/raiden-ts/src/transfers/epics/init.ts
+++ b/raiden-ts/src/transfers/epics/init.ts
@@ -22,7 +22,7 @@ import { RaidenEpicDeps } from '../../types';
 import { matrixPresence } from '../../transport/actions';
 import { getCap } from '../../transport/utils';
 import { Hash, untime } from '../../utils/types';
-import { distinctRecordValues, pluckDistinct } from '../../utils/rx';
+import { completeWith, distinctRecordValues, pluckDistinct } from '../../utils/rx';
 import {
   transferExpire,
   transferSigned,
@@ -213,6 +213,7 @@ export function transferClearCompletedEpic(
             startWith({ meta: pick(['secrethash', 'direction'], transfer) }),
           ),
         ),
+        completeWith(action$),
         // after some time with no action for this transfer going through (e.g. Processed retries)
         debounceTime(3 * httpTimeout),
         map(({ meta }) => transferClear(undefined, meta)), // clear transfer

--- a/raiden-ts/src/transfers/epics/locked.ts
+++ b/raiden-ts/src/transfers/epics/locked.ts
@@ -43,7 +43,7 @@ import { RaidenState } from '../../state';
 import { RaidenEpicDeps } from '../../types';
 import { isActionOf } from '../../utils/actions';
 import { LruCache } from '../../utils/lru';
-import { pluckDistinct } from '../../utils/rx';
+import { completeWith, pluckDistinct } from '../../utils/rx';
 import { Hash, Signed, UInt, Int, Address, untime, decode } from '../../utils/types';
 import { RaidenError, ErrorCodes } from '../../utils/error';
 import { Capabilities } from '../../constants';
@@ -1084,7 +1084,7 @@ export function transferGenerateAndSignEnvelopeMessageEpic(
   deps: RaidenEpicDeps,
 ) {
   const processedCache = new LruCache<string, Signed<Processed>>(32);
-  const state$ = deps.latest$.pipe(pluckDistinct('state')); // replayed(1)' state$
+  const state$ = deps.latest$.pipe(pluckDistinct('state'), completeWith(action$)); // replayed(1)' state$
   return merge(
     action$.pipe(
       filter(

--- a/raiden-ts/src/transfers/epics/secret.ts
+++ b/raiden-ts/src/transfers/epics/secret.ts
@@ -27,7 +27,7 @@ import { RaidenEpicDeps } from '../../types';
 import { isActionOf, isConfirmationResponseOf } from '../../utils/actions';
 import { RaidenError, ErrorCodes, assert, commonTxErrors } from '../../utils/error';
 import { fromEthersEvent, logToContractEvent } from '../../utils/ethers';
-import { pluckDistinct, retryWhile, takeIf } from '../../utils/rx';
+import { completeWith, pluckDistinct, retryWhile, takeIf } from '../../utils/rx';
 import { Hash, Secret, Signed, UInt, isntNil, untime } from '../../utils/types';
 import { getCap } from '../../transport/utils';
 import {
@@ -318,6 +318,7 @@ export function monitorSecretRegistryEpic(
         confirmationBlocks,
       ),
     ),
+    completeWith(state$),
     map(logToContractEvent<[Hash, Secret, Event]>(secretRegistryContract)),
     filter(isntNil),
     withLatestFrom(state$, config$),

--- a/raiden-ts/src/transfers/epics/utils.ts
+++ b/raiden-ts/src/transfers/epics/utils.ts
@@ -4,7 +4,7 @@ import { filter, ignoreElements, take } from 'rxjs/operators';
 
 import { messageSend } from '../../messages/actions';
 import { isResponseOf } from '../../utils/actions';
-import { repeatUntil } from '../../utils/rx';
+import { completeWith, repeatUntil } from '../../utils/rx';
 import { RaidenAction } from '../../actions';
 import {
   MessageType,
@@ -73,6 +73,7 @@ export function retrySendUntil$(
 ): Observable<messageSend.request> {
   return dispatchAndWait$(action$, send, isResponseOf(messageSend, send.meta)).pipe(
     repeatUntil(notifier, delayMs),
+    completeWith(action$),
   );
 }
 

--- a/raiden-ts/src/transport/epics/init.ts
+++ b/raiden-ts/src/transport/epics/init.ts
@@ -23,7 +23,6 @@ import {
   tap,
   toArray,
   mapTo,
-  finalize,
   first,
   timeout,
   pluck,
@@ -48,7 +47,7 @@ import { RaidenState } from '../../state';
 import { RAIDEN_DEVICE_ID } from '../../constants';
 import { getServerName } from '../../utils/matrix';
 import { decode } from '../../utils/types';
-import { pluckDistinct, retryWhile } from '../../utils/rx';
+import { completeWith, lastMap, pluckDistinct, retryWhile } from '../../utils/rx';
 import { matrixSetup } from '../actions';
 import { RaidenMatrixSetup } from '../state';
 import { Caps } from '../types';
@@ -396,6 +395,7 @@ export function initMatrixEpic(
         ),
       ),
     ),
+    completeWith(action$),
   );
 }
 
@@ -413,17 +413,16 @@ export function matrixShutdownEpic(
   {}: Observable<RaidenState>,
   { matrix$ }: RaidenEpicDeps,
 ): Observable<RaidenAction> {
-  return matrix$.pipe(
-    mergeMap((matrix) =>
-      action$.pipe(
-        finalize(async () => {
-          matrix.stopClient();
-          try {
-            await matrix.setPresence({ presence: 'offline', status_msg: '' });
-          } catch (err) {}
-        }),
-      ),
-    ),
-    ignoreElements(), // dont re-emit action$, but keep it subscribed so finalize works
+  return action$.pipe(
+    withLatestFrom(matrix$),
+    lastMap(async (pair) => {
+      if (!pair) return;
+      const matrix = pair[1];
+      matrix.stopClient();
+      try {
+        await matrix.setPresence({ presence: 'offline', status_msg: '' });
+      } catch (err) {}
+    }),
+    ignoreElements(),
   );
 }

--- a/raiden-ts/src/transport/epics/messages.ts
+++ b/raiden-ts/src/transport/epics/messages.ts
@@ -32,7 +32,7 @@ import constant from 'lodash/constant';
 
 import { EventType, MatrixClient, MatrixEvent, Room } from 'matrix-js-sdk';
 
-import { concatBuffer, pluckDistinct, retryWhile } from '../../utils/rx';
+import { completeWith, concatBuffer, pluckDistinct, retryWhile } from '../../utils/rx';
 import { intervalFromConfig, RaidenConfig } from '../../config';
 import { Capabilities } from '../../constants';
 import { Address, Signed } from '../../utils/types';
@@ -188,6 +188,9 @@ export function matrixMessageSendEpic(
               deps.log.error('messageSend error', err, actions[0].meta);
               return from(actions.map((action) => messageSend.failure(err, action.meta)));
             }),
+            // when shutting down, give up the 'wait' phase of sendAndWait$, but still give some
+            // time to concatBuffer to deplete buffer sending pending requests
+            completeWith(action$, 10),
           );
         }, 10),
       ),
@@ -217,17 +220,18 @@ function sendGlobalMessages(
     // send message!
     mergeMap(async (room) => {
       retries++;
-      return matrix.sendEvent(room.roomId, 'm.room.message', { body, msgtype: 'm.text' }, '');
+      return matrix
+        .sendEvent(room.roomId, 'm.room.message', { body, msgtype: 'm.text' }, '')
+        .then(() =>
+          log.info('messageGlobalSend success', {
+            tookMs: Date.now() - start,
+            retries,
+            roomName,
+            batchSize: actions.length,
+          }),
+        );
     }),
     retryWhile(intervalFromConfig(config$), { maxRetries: 3, onErrors: networkErrors }),
-    tap(() =>
-      log.info('messageGlobalSend success', {
-        tookMs: Date.now() - start,
-        retries,
-        roomName,
-        batchSize: actions.length,
-      }),
-    ),
     catchError((err) => {
       log.error('Error sending messages to global room', err, {
         retries,
@@ -265,6 +269,7 @@ export function matrixMessageGlobalSendEpic(
           return matrix$.pipe(
             withLatestFrom(config$),
             mergeMap(([matrix, config]) => sendGlobalMessages(actions, matrix, config, deps)),
+            completeWith(action$, 10),
           );
         }, 20),
       ),
@@ -304,7 +309,7 @@ function isValidMessage([{ matrix, event, room }, config]: [
  * @returns Observable of messageReceived actions
  */
 export function matrixMessageReceivedEpic(
-  {}: Observable<RaidenAction>,
+  action$: Observable<RaidenAction>,
   {}: Observable<RaidenState>,
   { log, matrix$, config$, latest$ }: RaidenEpicDeps,
 ): Observable<messageReceived> {
@@ -321,6 +326,7 @@ export function matrixMessageReceivedEpic(
         ),
       ),
     ),
+    completeWith(action$),
     withLatestFrom(config$),
     filter(isValidMessage),
     mergeMap(([{ event, room }, { httpTimeout }]) =>
@@ -331,6 +337,7 @@ export function matrixMessageReceivedEpic(
           const rooms = state.transport.rooms?.[presence.meta.address] ?? [];
           return !room || rooms.includes(room.roomId);
         }),
+        completeWith(action$),
         take(1),
         // take up to an arbitrary timeout to presence status for the sender
         // AND the room in which this message was sent to be in sender's address room queue

--- a/raiden-ts/src/transport/epics/presence.ts
+++ b/raiden-ts/src/transport/epics/presence.ts
@@ -33,7 +33,7 @@ import { RaidenEpicDeps } from '../../types';
 import { RaidenAction } from '../../actions';
 import { RaidenState } from '../../state';
 import { getUserPresence } from '../../utils/matrix';
-import { pluckDistinct, retryWhile } from '../../utils/rx';
+import { completeWith, pluckDistinct, retryWhile } from '../../utils/rx';
 import { matrixPresence } from '../actions';
 import { channelMonitored } from '../../channels/actions';
 import { parseCaps, stringifyCaps } from '../utils';
@@ -250,6 +250,7 @@ export function matrixPresenceUpdateEpic(
       // matrix's 'User.presence' sometimes fail to fire, but generic 'event' is always fired,
       // and User (retrieved via matrix.getUser) is up-to-date before 'event' emits
       fromEvent<MatrixPresenceEvent>(matrix, 'event').pipe(
+        completeWith(action$),
         filter((event) => event.getType() === 'm.presence'),
         // parse peer address from userId
         mergeMap(function* (event) {
@@ -274,6 +275,7 @@ export function matrixPresenceUpdateEpic(
     mergeMap((grouped$) =>
       grouped$.pipe(
         switchMap(([{ user, address, event }]) => fetchPresence$(user, address, event, deps)),
+        completeWith(action$),
       ),
     ),
     withLatestFrom(latest$),
@@ -323,11 +325,12 @@ export function matrixMonitorChannelPresenceEpic(
  * @returns Observable which never emits
  */
 export function matrixUpdateCapsEpic(
-  {}: Observable<RaidenAction>,
+  action$: Observable<RaidenAction>,
   {}: Observable<RaidenState>,
   { matrix$, config$ }: RaidenEpicDeps,
 ): Observable<never> {
   return config$.pipe(
+    completeWith(action$),
     pluck('caps'),
     distinctUntilChanged(isEqual),
     skip(1), // skip replay(1) and act only on changes

--- a/raiden-ts/src/transport/epics/rooms.ts
+++ b/raiden-ts/src/transport/epics/rooms.ts
@@ -30,7 +30,7 @@ import { RaidenState } from '../../state';
 import { channelMonitored } from '../../channels/actions';
 import { messageReceived } from '../../messages/actions';
 import { transferSigned } from '../../transfers/actions';
-import { pluckDistinct, retryWhile } from '../../utils/rx';
+import { completeWith, pluckDistinct, retryWhile } from '../../utils/rx';
 import { getServerName } from '../../utils/matrix';
 import { Direction } from '../../transfers/state';
 import { matrixRoom, matrixRoomLeave, matrixPresence } from '../actions';
@@ -148,6 +148,7 @@ export function matrixCreateRoomEpic(
           latest$.pipe(
             // wait for user to be monitored
             filter(({ presences }) => peer in presences),
+            completeWith(action$),
             take(1),
             // skip room creation/invite if both partner and us have ToDevice capability set
             filter(
@@ -214,6 +215,7 @@ export function matrixInviteEpic(
           if (!action.payload.available || !roomId) return EMPTY;
           return inviteLoop$(matrix, roomId, action.payload.userId, config, { log });
         }),
+        completeWith(action$),
       ),
     ),
   );
@@ -232,7 +234,7 @@ export function matrixInviteEpic(
  * @returns Observable of matrixRoom actions
  */
 export function matrixHandleInvitesEpic(
-  {}: Observable<RaidenAction>,
+  action$: Observable<RaidenAction>,
   {}: Observable<RaidenState>,
   { log, matrix$, config$, latest$ }: RaidenEpicDeps,
 ): Observable<matrixRoom> {
@@ -241,6 +243,7 @@ export function matrixHandleInvitesEpic(
     switchMap((matrix) =>
       fromEvent<[MatrixEvent, RoomMember]>(matrix, 'RoomMember.membership').pipe(
         map(([event, member]) => ({ event, member, matrix })),
+        completeWith(action$),
       ),
     ),
     filter(
@@ -256,6 +259,7 @@ export function matrixHandleInvitesEpic(
           Object.values(presences).find((p) => p.payload.userId === event.getSender()),
         ),
         filter(isntNil),
+        completeWith(action$),
         take(1),
         // Don't wait more than some arbitrary time for this sender presence update to show
         // up; completes without emitting anything otherwise, ending this pipeline.
@@ -353,6 +357,7 @@ export function matrixLeaveUnknownRoomsEpic(
         mergeMap(({ httpTimeout }) => timer(httpTimeout)),
       ),
     ),
+    completeWith(state$),
     withLatestFrom(state$, config$),
     // filter for leave events to us
     filter(([{ matrix, room }, { transport }, config]) => {
@@ -399,6 +404,7 @@ export function matrixCleanLeftRoomsEpic(
     switchMap((matrix) =>
       fromEvent<[Room, string]>(matrix, 'Room.myMembership').pipe(
         map(([room, membership]) => ({ room, membership, matrix })),
+        completeWith(state$),
       ),
     ),
     // filter for leave events to us

--- a/raiden-ts/src/transport/epics/webrtc.ts
+++ b/raiden-ts/src/transport/epics/webrtc.ts
@@ -46,7 +46,7 @@ import { MatrixClient } from 'matrix-js-sdk';
 import { Capabilities } from '../../constants';
 import { Address, decode, isntNil } from '../../utils/types';
 import { jsonParse, jsonStringify } from '../../utils/data';
-import { timeoutFirst } from '../../utils/rx';
+import { completeWith, timeoutFirst } from '../../utils/rx';
 import { RaidenEpicDeps } from '../../types';
 import { RaidenAction } from '../../actions';
 import { dispatchAndWait$, exponentialBackoff } from '../../transfers/epics/utils';
@@ -585,6 +585,7 @@ export function rtcConnectEpic(
             !!getCap(caps, Capabilities.WEBRTC),
         ),
         switchMap(([action, config]) => handlePresenceChange$(action, action$, config, deps)),
+        completeWith(action$),
       ),
     ),
   );

--- a/raiden-ts/src/utils/rx.ts
+++ b/raiden-ts/src/utils/rx.ts
@@ -25,6 +25,10 @@ import {
   mergeMapTo,
   tap,
   concatMap,
+  ignoreElements,
+  endWith,
+  last,
+  delay,
 } from 'rxjs/operators';
 import { isntNil } from './types';
 import { ErrorMatches, matchError } from './error';
@@ -222,7 +226,42 @@ export function takeIf<T>(
       takeUntil(distinctCond$.pipe(filter((cond): cond is false => !cond))),
       // re-subscribe input$ when cond becomes truty
       repeatWhen(() => distinctCond$.pipe(filter((cond): cond is true => cond))),
+      completeWith(input$),
     );
+}
+
+/**
+ * Complete an input when another observable completes
+ *
+ * @param complete$ - Observable which will complete input when completed
+ * @param delayMs - Delay completion by some time after complete$ completes
+ * @returns Operator returning observable mirroring input, but completes when complete$ completes
+ */
+export function completeWith<T>(
+  complete$: Observable<unknown>,
+  delayMs?: number,
+): MonoTypeOperatorFunction<T> {
+  return (input$) => {
+    let output$ = input$.pipe(takeUntil(complete$.pipe(ignoreElements(), endWith(null))));
+    if (delayMs !== undefined) output$ = output$.pipe(delay(delayMs));
+    return output$;
+  };
+}
+
+/**
+ * Like a mergeMap which only subscribes to the inner observable once the input completes;
+ * Intermediary values are ignored; project receives optionally the last value emitted by input,
+ * or null if no value was emitted
+ *
+ * @param project - callback to generate the inner observable, receives last emitted value or null
+ * @returns Operator returning observable mirroring inner observable
+ */
+export function lastMap<T, R>(
+  project: (lastValue: T | null) => ObservableInput<R>,
+): OperatorFunction<T, R> {
+  return (input$) => {
+    return input$.pipe(last(undefined, null), mergeMap(project));
+  };
 }
 
 /**

--- a/raiden-ts/tests/unit/epics/send.spec.ts
+++ b/raiden-ts/tests/unit/epics/send.spec.ts
@@ -116,6 +116,7 @@ describe('send transfer', () => {
       channel: channelUniqueKey(getChannel(raiden, partner)),
       secrethash,
     });
+    await sleep(100);
 
     expectChannelsAreInSync([raiden, partner]);
   });

--- a/raiden-ts/tests/unit/epics/transport.spec.ts
+++ b/raiden-ts/tests/unit/epics/transport.spec.ts
@@ -499,16 +499,16 @@ describe('matrixPresenceUpdateEpic', () => {
     const partnerMatrix = (await partner.deps.matrix$.toPromise()) as jest.Mocked<MatrixClient>;
 
     raiden.store.dispatch(matrixPresence.request(undefined, { address: partner.address }));
-    await sleep();
+    await sleep(raiden.config.httpTimeout);
     expect(raiden.output).toContainEqual(
       matrixPresence.success(
         expect.objectContaining({ available: true, userId: partnerMatrix.getUserId()! }),
         { address: partner.address },
       ),
     );
-    raiden.output = [];
+    raiden.output.splice(0, raiden.output.length);
 
-    matrix.getProfileInfo.mockRejectedValueOnce(new Error('could not get user profile'));
+    matrix.getProfileInfo.mockRejectedValue(new Error('could not get user profile'));
     matrix.emit('event', {
       getType: () => 'm.presence',
       getSender: () => partnerMatrix.getUserId(),
@@ -847,7 +847,7 @@ describe('matrixMessageSendEpic', () => {
       ),
     );
 
-    await sleep(raiden.config.httpTimeout);
+    await sleep(100);
     expect(raiden.output).toContainEqual(
       messageSend.success(expect.objectContaining({ via: expect.stringMatching(/^!/) }), {
         address: partner.address,
@@ -887,7 +887,7 @@ describe('matrixMessageSendEpic', () => {
       messageSend.request({ message }, { address: partner.address, msgId: message }),
     );
 
-    await sleep(3 * raiden.config.httpTimeout);
+    await sleep(200);
     expect(raiden.output).toContainEqual(
       messageSend.failure(expect.objectContaining({ message: 'Failed 4' }), {
         address: partner.address,

--- a/raiden-ts/tests/unit/mocks.ts
+++ b/raiden-ts/tests/unit/mocks.ts
@@ -1031,7 +1031,7 @@ export async function makeRaiden(
     async (n: string | number | Promise<string | number>) =>
       ({
         timestamp: Math.floor(
-          (Date.now() - (provider.blockNumber - +(await n)) * provider.pollingInterval) / 1000,
+          (Date.now() - (provider.blockNumber - +(await n) + 1) * provider.pollingInterval) / 1000,
         ),
       } as any),
   );
@@ -1299,7 +1299,6 @@ export async function makeRaiden(
     raidenReducer,
     initialState as any,
     applyMiddleware(
-      epicMiddleware,
       () => (next) => (action) => {
         // don't output before starting, so we can change state without side-effects
         if (raiden.started) {
@@ -1310,6 +1309,7 @@ export async function makeRaiden(
         }
         return next(action);
       },
+      epicMiddleware,
       createPersisterMiddleware(db),
     ),
   );
@@ -1336,8 +1336,9 @@ export async function makeRaiden(
       await synced;
     },
     started: undefined,
-    stop: () => {
+    stop: async () => {
       raiden.store.dispatch(raidenShutdown({ reason: ShutdownReason.STOP }));
+      await raiden.deps.db.busy$.toPromise();
       raiden.deps.provider.removeAllListeners();
       const idx = mockedClients.indexOf(raiden);
       if (idx >= 0) mockedClients.splice(idx, 1);

--- a/raiden-ts/tests/unit/utils.spec.ts
+++ b/raiden-ts/tests/unit/utils.spec.ts
@@ -1,7 +1,7 @@
 import * as t from 'io-ts';
 import { fold, isRight } from 'fp-ts/lib/Either';
 import { pipe } from 'fp-ts/lib/pipeable';
-import { first, mapTo, toArray } from 'rxjs/operators';
+import { delay, first, ignoreElements, mapTo, tap, toArray } from 'rxjs/operators';
 
 import { BigNumber } from '@ethersproject/bignumber';
 import { JsonRpcProvider, Web3Provider } from '@ethersproject/providers';
@@ -26,7 +26,7 @@ import { getLocksroot, makeSecret, getSecrethash } from 'raiden-ts/transfers/uti
 import { Lock } from 'raiden-ts/channels';
 import { LocksrootZero } from 'raiden-ts/constants';
 import { of, timer } from 'rxjs';
-import { concatBuffer } from 'raiden-ts/utils/rx';
+import { completeWith, concatBuffer, lastMap } from 'raiden-ts/utils/rx';
 import { getSortedAddresses } from 'raiden-ts/transport/utils';
 
 describe('getLogsByChunk$', () => {
@@ -392,4 +392,54 @@ test('getSortedAddresses', () => {
     '0x000000000000000000000000000000000000000A',
     '0x000000000000000000000000000000000000000b',
   ]);
+});
+
+test('completeWith', async () => {
+  expect.assertions(5);
+
+  const completeBefore = jest.fn();
+  const completeAfter = jest.fn();
+  await expect(
+    timer(20)
+      .pipe(
+        tap({ complete: completeBefore }),
+        completeWith(timer(10)),
+        tap({ complete: completeAfter }),
+      )
+      .toPromise(),
+  ).resolves.toBeUndefined();
+  // since completeWith completed the output before the input timer, it seems like it was
+  // unsubscribed instead, but second tap sees proper completion
+  expect(completeBefore).not.toHaveBeenCalled();
+  expect(completeAfter).toHaveBeenCalled();
+
+  // since completeWith emits synchronously, it should unsubscribe from (even synchronous) input
+  // before first emition
+  await expect(
+    of(1)
+      .pipe(completeWith(of(2)))
+      .toPromise(),
+  ).resolves.toBeUndefined();
+
+  // timer(0) is immediate but asynchronous, synchronous input should still go through
+  await expect(
+    of(1)
+      .pipe(completeWith(timer(0)))
+      .toPromise(),
+  ).resolves.toBe(1);
+});
+
+test('lastMap', async () => {
+  const project = jest.fn(async () => true);
+  await expect(of(1, 2, 3).pipe(lastMap(project)).toPromise()).resolves.toBe(true);
+  expect(project).toHaveBeenCalledTimes(1); // only last emition passes
+  expect(project).toHaveBeenCalledWith(3, expect.anything());
+
+  const project2 = jest.fn((v) => of([v]).pipe(delay(20)));
+  await expect(timer(10).pipe(ignoreElements(), lastMap(project2)).toPromise()).resolves.toEqual([
+    null,
+  ]);
+  expect(project2).toHaveBeenCalledTimes(1); // only last emition passed
+  // ignoreElements calls project with null
+  expect(project2).toHaveBeenCalledWith(null, expect.anything());
 });

